### PR TITLE
added hw-8-1

### PR DESCRIPTION
**Pull request by kenu21**

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,48 +3,170 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static class Node<T> {
+        T item;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: add access modifier

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,48 +3,170 @@
...
+        MyLinkedList.Node<T> next;
+        MyLinkedList.Node<T> prev;
+
+        Node(MyLinkedList.Node<T> prev, T element, MyLinkedList.Node<T> next) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: the same
```suggestion
        Node(Node<T> prev, T element, Node<T> next) {
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,48 +3,170 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static class Node<T> {
+        T item;
+        MyLinkedList.Node<T> next;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
        Node<T> next;
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,48 +3,170 @@
...
+
+        Node(MyLinkedList.Node<T> prev, T element, MyLinkedList.Node<T> next) {
+            item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private int size;
+    private MyLinkedList.Node<T> first;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
    private Node<T> first;
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,48 +3,170 @@
...
+    private MyLinkedList.Node<T> last;
+
+    public MyLinkedList() {
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        MyLinkedList.Node<T> l = last;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: СS

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,48 +3,170 @@
...
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            MyLinkedList.Node<T> l = last;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: maybe better call method add(T value)?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 7]
```diff
@@ -3,48 +3,170 @@
...
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        MyLinkedList.Node<T> x = node(index);
+        T oldVal = x.item;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: do we need this variable?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 8]
```diff
@@ -3,48 +3,170 @@
...
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        return unlink(node(index));
     }
 
     @Override
     public T remove(T t) {
+        MyLinkedList.Node x;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: move Node x in the for loop

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 9]
```diff
@@ -3,48 +3,170 @@
...
+        if (t == null) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: can we use one for loop?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 10]
```diff
@@ -3,48 +3,170 @@
...
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private MyLinkedList.Node<T> node(int index) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
    private MyLinkedList.Node<T> getNode(int index) {
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 11]
```diff
@@ -3,48 +3,170 @@
...
+        MyLinkedList.Node x;
+        int i;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: the same

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 12]
```diff
@@ -3,48 +3,170 @@
...
+        } else {
+            x = last;
+            for (i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
+    T unlink(MyLinkedList.Node<T> x) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: CS

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 13]
```diff
@@ -3,48 +3,170 @@
...
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private MyLinkedList.Node<T> node(int index) {
+        MyLinkedList.Node x;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: CS

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 14]
```diff
@@ -3,48 +3,170 @@
...
+            prev.next = next;
+            x.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            x.next = null;
+        }
+        x.item = null;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: we can write:
node = null
and then remove  x.prev = null; and x.next = null;

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 6 above)
Comment (kenu21) [reply src/main/java/core/basesyntax/MyLinkedList.java]: Yes, I'll fix it.

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 7 above)
Comment (kenu21) [reply src/main/java/core/basesyntax/MyLinkedList.java]: No. I'll fix it.

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 9 above)
Comment (kenu21) [reply src/main/java/core/basesyntax/MyLinkedList.java]: Yes, we can. I'll do it.

**Pull request closed without merging**